### PR TITLE
Polyfill the incremental generator ForAttributeWithMetadataName from roslyn (for EventSourceGeneration).

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/runtime/pull/70911 and https://github.com/dotnet/runtime/pull/71652

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs [hunk 1]
```diff
@@ -4,116 +4,90 @@
...
+                        // since this generator doesn't know how to generate a nested type...
                         continue;
                     }
+                }
 
-                    if (!attributeFullName.Equals(EventSourceAttribute, StringComparison.Ordinal))
-                    {
-                        continue;
-                    }
+                if (attribute.AttributeClass.Name != "EventSourceAttribute" &&
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:41]: Is `&&` correct here?  It shouldn't be `||`?

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs [hunk 2]
```diff
@@ -61,19 +62,13 @@ public static class StepNames
...
-                        MethodDeclarationSyntax syntax = (MethodDeclarationSyntax)context.Node;
-                        if (context.SemanticModel.GetDeclaredSymbol(syntax, ct) is IMethodSymbol methodSymbol
-                            && methodSymbol.GetAttributes().Any(static attribute => attribute.AttributeClass?.ToDisplayString() == TypeNames.LibraryImportAttribute))
-                        {
-                            return new { Syntax = syntax, Symbol = methodSymbol };
-                        }
-
-                        return null;
-                    })
+                .ForAttributeWithMetadataName(
```
Comment (stephentoub) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs:65]: Is it intentional that LibraryImportGenerator is being modified in this PR?

Comment (stephentoub) [reply src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs:65]: @CyrusNajmabadi, can you remove the LibraryImport changes from this PR?  I'm assuming they're erroneous (and they're causing test failures).

**Comment by CyrusNajmabadi** [comment]
This is a follow-up pr to the library import pr. @joperezr tells me that this isn't shipping yet, so it was low pri to fix.

**Comment by stephentoub** [comment]
> This is a follow-up pr to the library import pr.

Ok.  Out of curiosity why does this depend on the LibraryImport one?

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs]: Done.

File: src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs [hunk 3]
```diff
@@ -4,119 +4,96 @@
...
 {
     public partial class EventSourceGenerator
     {
-        private static bool IsSyntaxTargetForGeneration(SyntaxNode node, CancellationToken cancellationToken) =>
-            node is ClassDeclarationSyntax { AttributeLists.Count: > 0 };
-
-        private static EventSourceClass? GetSemanticTargetForGeneration(GeneratorSyntaxContext context, CancellationToken cancellationToken)
+        private static EventSourceClass? GetSemanticTargetForGeneration(GeneratorAttributeSyntaxContext context, CancellationToken cancellationToken)
         {
-            const string EventSourceAutoGenerateAttribute = "System.Diagnostics.Tracing.EventSourceAutoGenerateAttribute";
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:24]: we don't need to check for EventSourceAutoGenerateAttribute.  we wouldn't have gotten here unless we had that attribute on the class we're examining.

File: src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs [hunk 4]
```diff
@@ -4,119 +4,96 @@
...
+                    continue;
+                }
 
-                    string attributeFullName = caSymbol.ContainingType.ToDisplayString();
+                nspace ??= ConstructNamespace(ns);
 
-                    if (attributeFullName.Equals(EventSourceAutoGenerateAttribute, StringComparison.Ordinal))
-                    {
-                        autoGenerate = true;
-                        continue;
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:47]: we required having EventSourceAutoGenerateAttribute before.  otherwise, autogenerate would not be true, and we'd bail out below.

File: src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs [hunk 5]
```diff
@@ -4,119 +4,96 @@
...
-                    }
+                string className = classDef.Identifier.ValueText;
+                string name = className;
+                string guid = "";
 
-                    if (!attributeFullName.Equals(EventSourceAttribute, StringComparison.Ordinal))
-                    {
-                        continue;
-                    }
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:53]: the code below would only execute if we had EventSourceAttribute, so that's what we maintain below.

File: src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs [hunk 6]
```diff
@@ -4,119 +4,96 @@
...
+            var classDef = (ClassDeclarationSyntax)context.TargetNode;
+            NamespaceDeclarationSyntax? ns = classDef.Parent as NamespaceDeclarationSyntax;
+            if (ns is null)
+            {
+                if (classDef.Parent is not CompilationUnitSyntax)
+                {
+                    // since this generator doesn't know how to generate a nested type...
+                    return null;
+                }
+            }
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:31]: moved this check up.  there was no point doing it in a loop since it would always be the same result.  and if it always caused us to continue, we'd never set `eventSourceClass`.  so we'd terminate the loop and would return `null`.  so this is effectively the same.

Comment (CyrusNajmabadi) [reply src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:31]: note: this logic is wrong in that it can't handle:  `namespace n1 { namespace n2 { class Whatever { } } }`.  However, that bug existed before.  I didn't change it.  

A better check would likely be to do `if (classDef.Parent is TypeDeclarationSyntax)`

File: src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs [hunk 7]
```diff
@@ -4,119 +4,96 @@
...
+
             EventSourceClass? eventSourceClass = null;
+            string? nspace = null;
 
-            bool autoGenerate = false;
-            foreach (AttributeListSyntax cal in classDef.AttributeLists)
+            foreach (AttributeData attribute in context.TargetSymbol.GetAttributes())
             {
-                foreach (AttributeSyntax ca in cal.Attributes)
+                if (attribute.AttributeClass?.Name != "EventSourceAttribute" ||
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:38]: in general, try to do the basic name check first.  it's much faster and allocation-light versus ToDisplayString

File: src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs [hunk 8]
```diff
@@ -4,119 +4,96 @@
...
-                    continue;
-                }
-            }
+        private static string? ConstructNamespace(NamespaceDeclarationSyntax? ns)
+        {
+            if (ns is null)
+                return string.Empty;
 
-            if (!autoGenerate)
+            string nspace = ns.Name.ToString();
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs:84]: note.  this is not correct.  if you have `namespace N1/*comment*/. N2`, then the ToString will include hte extraneous comments and whitespace.

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs [hunk 9]
```diff
@@ -358,13 +358,13 @@ private static MarshallingGeneratorFactoryKey<(TargetFramework, Version, Library
...
                     new CharMarshallingGeneratorFactory(generatorFactory, useBlittableMarshallerForUtf16: true),
-                    new AttributedMarshallingModelOptions(runtimeMarshallingDisabled, Scenario.ElementIn, Scenario.ElementRef, Scenario.ElementOut));
+                    new AttributedMarshallingModelOptions(runtimeMarshallingDisabled, MarshalMode.ElementIn, MarshalMode.ElementRef, MarshalMode.ElementOut));
                 // We don't need to include the later generator factories for collection elements
                 // as the later generator factories only apply to parameters.
                 generatorFactory = new AttributedMarshallingModelGeneratorFactory(
                     generatorFactory,
                     elementFactory,
-                    new AttributedMarshallingModelOptions(runtimeMarshallingDisabled, Scenario.ManagedToUnmanagedIn, Scenario.ManagedToUnmanagedRef, Scenario.ManagedToUnmanagedOut));
+                    new AttributedMarshallingModelOptions(runtimeMarshallingDisabled, MarshalMode.ManagedToUnmanagedIn, MarshalMode.ManagedToUnmanagedRef, MarshalMode.ManagedToUnmanagedOut));
```
Comment (stephentoub) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs:367]: I'm still not clear on why these changes are necessary in this PR.  Is this PR renaming Scenario to MarshalMode or something?

**Review by stephentoub — APPROVED** [review]
Thanks!

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs (hunk 9 above)
Comment (stephentoub) [reply src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs]: > If you want though i can pre-squash.

Nope.
